### DISCUSSION TITLE
Groundwork for porting to other platforms/chipsets

### DISCUSSION THREAD

**Discussion by ghost** [selfcustody/krux#170] [Ideas]
Related:
https://github.com/selfcustody/krux/issues/39
https://github.com/selfcustody/krux/issues/123
https://github.com/selfcustody/krux/issues/161

This discussion is meant to tie together the various issues I've added in the past months and offer a more coherent vision of where I think we should try to take Krux.

I would like Krux to become something like the Linux (or perhaps Ubuntu is a better analogue) of hardware wallet firmware, an open-source device-agnostic signing "OS" you can install on any platform capable of running Micropython. Thanks to @odudex's work, we've expanded the K210-based devices that it supports, but I hope to eventually shim out the MaixPy/K210-specific modules so we can port to other platforms and not be too locked-in to any one vendor. It would be cool to reach a point where someone can come up with their own set of {board + camera + sd card reader + lcd} and either have it "Just Work" or require a small amount of dev effort to wire up their hardware.

When I made the Krux Simulator, I had to create mocks of every module that was not available under Python 3, which can be seen here:
https://github.com/selfcustody/krux/tree/develop/simulator/kruxsim/mocks

This will be a useful reference of what it should take to port Krux to other boards including ESP32, STM32, etc.

One important thing to note is that MaixPy uses OpenMV which adds the `sensor`, `lcd`, and `image` modules that we use. Looking more closely at micropython, it appears they don't include baked-in camera support. Furthermore, the [list of supported cameras](https://github.com/openmv/openmv/tree/master/src/omv/sensors) in OpenMV is large and the project seems to be actively developed, so I think we should continue using OpenMV for now. This may mean that we should fork the [openmv repo](https://github.com/openmv/openmv/) and use it as the basis for our firmware. One big gotcha with OpenMV, however, is that I believe it (currently) doesn't support ESP32 devices and is instead focused on STM32. But we can cross that bridge when we get to it.

First things first, we should research and select a non-K210 board to target based on known compatibility with micropython and OpenMV and specs that would support running Krux (e.g., 2MB+ flash without firmware upgrade support, 4MB+ flash for firmware upgrade support). I'm leaning toward the [OpenMV Cam H7 R2](https://www.sparkfun.com/products/18982) with an [LCD shield](https://www.sparkfun.com/products/16777). 

What do you think, @odudex? Any better options or other input on this?

**Comment by odudex** [comment]
@jreesun you read my mind on this issue. I agree this is an important move, as we run on four devices today, but rely on a single chip model, the K210. Cam H7 R2 seems to be the perfect target for next port.
STM32H743II with Cortex M7@480 MHz has the horsepower required for smooth video applications, ST is widely used, relatively cheap and don't having integrated radio (Wifi, BLE, NFC) is also an advantage for secure devices.
ESP32s @240MHz may lead to poor performance or a hard to achieve level of optimization. (Seedsigner devs will tell if I'm right)
Maybe when Espressif launch their Risc-V chips Krux could incorporate them.
Maixpy looks to me poorly documented version of OpenMV adapted for Sipeed's hardware. Migrate to OpenMV would mean incorporate a better structured and maintained project, but the challenge will be to "merge" Sipeed's hardware (k210 tool-chain, cams and displays) smoothly to it.

**Reply by odudex** [reply]
I also wonder, how hard would it be to compile a Krux android app (for learn and test purposes)? I think that, starting from your work with the simulator, build an app with something like Kivy could be interesting.

**Reply by ghost** [reply]
I also wasn't sure about the feasibility of ESP32, but the potential devices you can run on was appealing. Will be interesting to see what SeedSigner devs can do.

Kivy is a great idea! It looks like you can build for virtually every platform, so if we did it right we could even replace the simulator with it.

Phones make a lot of sense as a target if they were able to be used in a secure way. They can be quite cheap, have good cameras and displays, etc. If they exposed their secure elements via an API we could call to store a key or have a signature created, I would feel more comfortable using them in that way. Though, we would need to make some changes to Krux to support off-loading all key management to an external API.

Regardless, some users may have old phones (or computers) they use as offline key stores that could take advantage of Krux if we supported it, so I'm all for exploring this avenue!

**Comment by odudex** [comment]
Your child [on Android](https://github.com/odudex/krux_binaries/tree/main/Android):
Working on a Kivy-Buildozer setup, it's functional now

**Reply by ghost** [reply]
@odudex I see you've been adding a lot of new features! Kivy, Nostr, Tinyseed, battery life indicator, etc. Very cool :) Any plans to turn them into PR(s) soon-ish? Asking because I was thinking about cutting a new release shortly so the changes you already made in `develop` could finally come out--particularly the compact seed QR work--but I don't want to do that if you plan to add new work soon. No pressure, just want to coordinate a bit!

Speaking of which, which issues in this list have you already worked on?
https://github.com/selfcustody/krux/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Next+Major+Release%22

Also, I've noticed a nice uptick in traffic and downloads over the past several months; Krux officially crossed the 1k downloads milestone! Exciting stuff.

**Reply by odudex** [reply]
Hi Jeff, nice to hear from you!
Krux have been a hobby for me. The problem is I have been undisciplined, doing what pleases me: Adding features. But I have not been covering them on pytest.
You can see [here](https://github.com/odudex/krux/tree/odudex_develop) that there are 62 commits, some from issue list, and many other things.
I did not publish the Kivy Android setup yet. Beyond mocks, it has some code customization in src and also libraries. My goal is to have something neatly merged, as you did with pygame simulator, but I don't know if I will be able to achieve it alone, may need some help.

From the issues list I committed on:
- Removed dice rolls restriction
- Amigo printer support (and new print options)
- Battery indicator

Now there's a ongoing work on:
- Faster thermal prints (driver improvement)

Some ideas/demands that are in my roadmap
- Show receive addresses
- Create Legacy script/derivation wallets
- Create custom derivations
- QR passphrases
- New helpers to transcript QRs (highlight regions, lines)

-- Nostr
- Proper (standard?) key derivations
- Import JSON events (not only event ID)
- NIP26 delegation(waiting standards to settle)

-- App
- Save encrypted wallet (app only)

If you tell me you're returning to the ship, I'll prioritize your demands. What should I focus on to have my commits merged? There's so much maybe we should beak them in sub-tasks/milestones to review and cover tests

**Reply by ghost** [reply]
In terms of what makes sense to add now, I think the first three you listed make the most sense if you've already done the work:

    - Removed dice rolls restriction
    - Amigo printer support (and new print options)
    - Battery indicator

Every commit in that branch you linked looked reasonable to include as long as the code is working.

Ideally, each feature would be its own PR, but if they're hard to separate or depend on prior commits, I don't mind going over a longer PR to get caught up again.

Is the Tinyseed work done as well, or is that ongoing? That's the only one I'd really prefer be in its own branch/PR since I imagine it has a lot more code.

Don't worry about adding test coverage for now, I can try to help with that after the PR(s) are reviewed so I don't slow you down too much.

I think the Nostr work will need to be discussed so we can nail down how best to integrate it and what all to integrate. I think it's worthwhile to add the ability to sign a Nostr post, but I'm not sure I want to add a bunch of Nostr-specific code since that would be outside the scope of this project. Ideally, you would be able to follow the existing flow of importing a mnemonic (perhaps with a [custom derivation](https://github.com/selfcustody/krux/issues/103)?) as well as the existing Sign Message functionality in order to sign a post. If it ends up feeling clunky or awkward to add support for Nostr, maybe there could be a fork of Krux that focuses exclusively on Nostr rather than Bitcoin. Regardless, it'll require some thought and discussion.

**Comment by odudex** [comment]
As far as I remember, the Tinyseed and Stackbit commits and files are easy to isolate. I'll clean the code and try to do a BIG PR this week with everything except tiny seed and Nostr.
Tiny seed import/export/print work and even the scan is working. I encourage to test it with my binaries by scanning a thermal printed tiny seed.
Sure nostr must be discussed, let's talk about this later. Just a fun fact, first air gapped post in the world was done using Krux.[Video](https://twitter.com/odudex/status/1611179199724933120)

**Comment by odudex** [comment]
Added code and instructions to build Krux Android app on "android" branch of my fork. I think it works quite well, but could be integrated on the project in a more neat way.

https://github.com/odudex/krux/blob/android/android/README.md

**Reply by ghost** [reply]
I made an issue for your Kivy work here: https://github.com/selfcustody/krux/issues/193

PS: Looking at prior release dates of Krux, it seems that roughly every 6 months we're ready for a new one. So, I've decided to make that official from here on out and follow a twice-a-year release schedule, with the soonest upcoming release being in April (23.04.0), and the following release in October (23.10.0). Kind of a weird, happy accident that we landed on the same months as Ubuntu.

I added two milestones for those releases that you can see here: https://github.com/selfcustody/krux/milestones

Your Kivy work is under the 23.10.0 milestone.

**Reply by odudex** [reply]
Nice! Seems a very effective way to plan next features.

**Comment by odudex** [comment]
Ps: The app contains the Nostr and "Seed QR regions transcript helper" experimental features. But if we have new official releases I plan to match the app code with main code.

**Reply by ghost** [reply]
Thanks! The regions stuff is pretty cool. Is it stable? I was planning to keep it in.

**Comment by odudex** [comment]
The way I did would require a change in Maixpy. I plan to re-make with rectangles so no Maixpy change is required
